### PR TITLE
async

### PR DESCRIPTION
**Pull request by durangatan**

### CODE REVIEW

**Comment by rquadling** [comment]
Hi. Doesn't the archive option need to be extended to include all the JS files into the archive?

```
data "archive_file" "sqs_lambda" {
  type        = "zip"
  source_dir = "${path.module}/lambda"
  output_path = "${path.module}/lambda_function.zip"
}
```

`source_dir` rather than `source_file`?

**Comment by cdornsife** [comment]
index.js is the only file in the lambda directory at the moment.

**Comment by rquadling** [comment]
Not in this PR it isn't. :-D

File: lambda/index.js [hunk 1]
```diff
@@ -1,92 +1,74 @@
...
 
-    });
+const handleInvalidationError = async err => {
+  if (err.code !== "TooManyInvalidationsInProgress") {
+    const warning = `ignoring error: ${err}.`;
+    logWarning(warning);
+    return Promise.resolve();
+  }
+
+  if (shouldRetry(message)) {
```
Comment (rquadling) [inline lambda/index.js:21]: message is not supplied to handleInvalidationError

File: lambda/index.js [hunk 2]
```diff
@@ -1,92 +1,74 @@
...
+    }
+
+    try {
+      const successfulInvalidation = await createCloudfrontInvalidation(
+        makeInvalidationParams(message)
+      );
+      log(successfulInvalidation);
+      return Promise.resolve();
+    } catch (e) {
+      return await handleInvalidationError(e);
```
Comment (rquadling) [inline lambda/index.js:70]: Should either supply message or event or record (I'm not an expert on JS, so sort of guessing which is appropriate) and have handleInvalidationError handle it appropriately.

File: lambda/aws.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      Paths: {
+        Quantity: 1,
+        Items: [message.path]
+      }
+    }
+  };
+}
+export function getQueueUrlFromArn(arn) {
+  const arnParts = arn.split(":", 6);
+  return `https://sqs.${arnParts[3]}.amazonaws.com/${arnParts[4]}/${
```
Comment (rquadling) [inline lambda/aws.js:36]: I think the line split here may not be suitable.

**Comment by cdornsife** [comment]
Oh, good point. The person who wrote this never actually ran it. I told him to PR what he had as is.

**Comment by rquadling** [comment]
I'm not a JS expert so my comments need to be ... 'serving suggestions' at best.

One thing I'd like to know is does the lambda function have to wait for the completion of the invalidation? Or can it be a "fire and forget"?

I've got the current master branch working great for me.

One thing though, can you do a release of the current master. That way, I can bake the version into my setup and not have to worry about master changing and my deployments suddenly not working.

**Comment by cdornsife** [comment]
The current master is missing some vital settings. I submitted a PR today that fixes it. The plan is to create a branch for a compatible terraform version like "0.11.8" then you can ?ref=0.11.8
The validation is async from what I can remember.

**Pull request closed without merging**